File: JavaCheatSheet/src/main/java/com/github/jameswdelancey/javaCheatSheet/Part12.java
package com.github.jameswdelancey.javaCheatSheet;

import java.util.ArrayList;
import java.util.List;

class Part12 {
    static class SomeHugeType {
        int[] data = new int[10_000_000];
    }

    static SomeHugeType allocateNumbers() {
        SomeHugeType numbers = new SomeHugeType();
        return numbers;
    }

    static SomeHugeType generateNumbersOnHeap() {
        SomeHugeType numbers = allocateNumbers();
        for (int i = 0; i < numbers.data.length; i++) {
            numbers.data[i] = i * 2;
        }
        return numbers;
    }

    static final int ANSWER = 42;
    static String QUESTION = "";

    static SomeHugeType generateNumbers() {
        SomeHugeType numbers = new SomeHugeType();
        for (int i = 0; i < numbers.data.length; i++) {
            numbers.data[i] = i * 2;
        }
        return numbers;
    }

    public static void main(String[] args) {
        generateNumbers();
    }
}

class Part12_2 {
    public static void main(String[] args) {
        Part12.generateNumbersOnHeap();
    }
}

class Part12_3 {
    static class Node {
        int value;
        Node left;
        Node right;

        Node(int value) {
            this.value = value;
        }

        void display() {
            System.out.println(value);
            if (left != null) {
                left.display();
            }
            if (right != null) {
                right.display();
            }
            System.out.println("UP");
        }
    }

    public static void main(String[] args) {
        Node root = new Node(2);
        root.left = new Node(1);
        root.right = new Node(3);
        root.display();
    }
}

class Part12_4 {
    interface Animal {
        void speak();
    }

    static class Cat implements Animal {
        @Override
        public void speak() {
            System.out.println("Meow!");
        }
    }

    static class Dog implements Animal {
        @Override
        public void speak() {
            System.out.println("Woof!");
        }
    }

    static void workWithAnimals(List<Animal> animals) {
        for (Animal animal : animals) {
            animal.speak();
        }
    }

    public static void main(String[] args) {
        Cat cat = new Cat();
        Dog dog = new Dog();
        List<Animal> animals = new ArrayList<>();
        animals.add(cat);
        animals.add(dog);
        workWithAnimals(animals);
    }
}

File: rustCheatSheet/src/part12.rs
use std::alloc::{alloc, Layout};
fn allocateNumbers() -> Box<SomeHugeType> {
    let layout = Layout::new::<SomeHugeType>();
    let ptr = unsafe { alloc(layout) } as *mut SomeHugeType;
    unsafe { Box::from_raw(ptr) }
}
fn generateNumbersOnHeap() -> Box<SomeHugeType> {
    let mut numbers = allocateNumbers();
    for (i, n) in numbers.iter_mut().enumerate() {
        *n = (i * 2) as i32;
    }
    numbers
}
const ANSWER: i32 = 42;
static mut QUESTION: String = String::new();

type SomeHugeType = [i32; 10_000_000];

fn generateNumbers() -> SomeHugeType {
    let mut numbers = [0; 10_000_000];
    for (i, n) in numbers.iter_mut().enumerate() {
        *n = (i * 2) as i32;
    }
    numbers
}

#[test]
fn main() {
    generateNumbers();
}

#[test]
fn main2() {
    generateNumbersOnHeap();
}

struct Node {
    value: i32,
    left: Option<Box<Node>>,
    right: Option<Box<Node>>,
}

impl Node {
    fn display(&self) {
        println!("{}", self.value);
        if let Some(ref left) = self.left {
            left.display();
        }
        if let Some(ref right) = self.right {
            right.display();
        }
        println!("UP");
    }
}

#[test]
fn main3() {
    let root = Node {
        value: 2,
        left: Some(Box::new(Node {
            value: 1,
            left: None,
            right: None,
        })),
        right: Some(Box::new(Node {
            value: 3,
            left: None,
            right: None,
        })),
    };
    root.display();
}

struct Cat {}
struct Dog {}
trait Animal {
    fn speak(&self);
}
impl Animal for Cat {
    fn speak(&self) {
        println!("Meow!");
    }
}
impl Animal for Dog {
    fn speak(&self) {
        println!("Woof!");
    }
}
fn workWithAnimals(animals: &[Box<dyn Animal>]) {
    for animal in animals {
        animal.speak();
    }
}
#[test]
fn main4() {
    let cat = Cat {};
    let dog = Dog {};
    let mut animals: Vec<Box<dyn Animal>> = Vec::new();
    animals.push(Box::new(cat));
    animals.push(Box::new(dog));
    workWithAnimals(&animals);
}
